### PR TITLE
Switch AWT font loading to Skia-based

### PR DESCRIPTION
**Pull request by rock3r**
This PR does many things, but for one main reason: improve font loading in Skiko on AWT (and thus, Compose Desktop).

## Improve font loading and listing
Available font listing doesn't rely anymore on discovering and loading font files from the disk. This is MUCH faster than the previous solution, which could take seconds on a performant machine. Besides, this greatly reduces memory usage as we're only holding the font family names list in memory, and cache items in memory lazily only once they are requested to speed up later lookups. Users can manually clear the memory cache, and register custom fonts (from files or resources).

We now get the list of available font families directly from Skia, with some additional features such as adding entries for the system font on macOS (which is hidden by default).

To note, the list of fonts available to Skia is not the same as available to AWT, for a number of reasons:
 1. AWT adds logical fonts (e.g., `Dialog`)
 2. AWT silently substitutes broken fonts
 3. AWT lists the macOS system font, but under a weird name, `.AppleSystemUIFont`, whereas Skia doesn't
 4. AWT and Skia don't use the same logic to list installed fonts on all platforms
 5. AWT includes fonts embedded into the JVM automatically

This makes things a bit more complicated, but if you _know_ the correct family name, everything works fine. While AWT silently substitutes invalid font family names with others, Skia will return `null` if it can't find a font. We follow the Skia approach, returning null if we can't match a font family name.

## Improve AWT `Font` conversion to Skiko `Typeface`
On the JetBrains Runtime, the `Font.toSkikoTypefaceOrNull()` will be able to properly match an AWT `Font` to a Skiko `Typeface`, even on OSes other than macOS. If running on other JVMs, this returns `null`, as requested by @igordmn. This requirement is documented in the KDocs.

The necessity to be on the JBR is due to the "regular" AWT implementations of `sun.font.Font2D` and `sun.funt.TrueTypeFont` being essentially broken: they read the wrong entry from the `name` table in TrueType fonts and use it as font family name. In the JBR, additional logic has been added to read the correct entry, the [_preferred family_](https://developer.apple.com/fonts/TrueType-Reference-Manual/RM06/Chap6name.html), and we can use this information to match with Skia font families, as Skia uses the correct _preferred family name_.

## Misc changes
* Cleanup codebase
* Added `@Ignore` to a number of failing tests

### CODE REVIEW

**Comment by rock3r** [comment]
@igordmn @lamba92 I pushed a commit where I upgraded the Coroutines version and fixed all tests that got broken as a result of it (I needed the newer Coroutines version for a fix for the AWT interop tests).

One of the tests, `FrameLimiterTest`, is broken and I can't fix it, as I don't know how to implement the custom scheduler wizardry with the new test API/behaviour. I have tried for a couple days now figuring out how to do it, but it seems like the _very_ deprecated `runBlockingTest` API it's using has somehow changed its behaviour. I tried converting it to the newer, and more functionally sound, `runTest` API, but then I have the issue that there doesn't seem to be any way to create a custom `TestDispatcher` (abstract class with internal ctor in coroutines-test), nor `TestCoroutineScheduler` (closed class), so I cannot emulate the imprecise scheduling of delays in any obvious way.

Folks, if you could take care of this, it'd be great, because I can't waste any more time trying to get this to work :(

The only changes I made to this class are minor cleanups, and things suggested to me on the Kotlin Slack by Dima Khalanskiy, which also don't seem to help much :(

**Comment by igordmn** [comment]
> Folks, if you could take care of this, it'd be great, because I can't waste any more time trying to get this to work

Let's just add @Ignore to these tests for now, and when you PR is ready for review, I will look at it, and will try to properly fix it.

**Comment by rock3r** [comment]
Some updates on this...

✔️   This code works 100% on macOS
❌   This code works fine on Windows, too — but AWT doesn't, so we need to figure out how to properly match things up there. I am studying the IJ codebase to see how they handle these issues (and I found and reported bugs there, too)
⏳   I haven't managed to get the Skiko project to compile on Linux yet, despite trying in a number of ways. I have shelved it while I work on adjusting things to work on Windows, as it'll require another test run anyway.

**Comment by rock3r** [comment]
Rebased on the main branch, and implemented the results of the past week of profuse swearing. There are good news, and bad news.

The bad news is that there isn't much we can do to fix the default AWT's broken font family enumeration logic on WIndows (and, theoretically, Linux, although I still haven't managed to get a VM working to compile and test there).

The good news is that, when we run on the JBR, we can and now do provide the correct behaviour on all platforms. This is really the most we can hope for, at this point, since doing otherwise would mean not relying on AWT at all on those other JVMs/OSes, which is kind of silly since that's the whole point here.

On macOS, the JVM behaves correctly by virtue of the platform-specific part of the code in `sun.awt` and `sun.font` simply asking the native APIs and not trying to do any parsing by themselves. On Windows and Linux, when running on the JBR, we can access some internal APIs in `sun.font` that have been fixed by the angels over at the JBR team. This behaviour is modeled on what IJ does in its UI, in the [`Settings > Editor > Font`](jetbrains://Idea/settings?name=Editor--Font) page, which also works on Windows and Linux. I have also added some logic to make sure that if the AWT logic fonts (e.g., _Dialog_) are used, we can resolve the actual physical underlying fonts and use those.

There is one small hack involved to make the JBR-specific behaviour happen: using ByteBuddy, I obtain access to the Java instrumentation APIs to override the encapsulation settings for the `sun.font` package in the `java.desktop` module. This is necessary because that package would be otherwise inaccessible, even by reflection. The only alternative is to pass an `--add-open` parameter to the `java` executable whenever running anything that uses Skiko, but it's a terrible user experience and we don't want to force users to do this.

**Comment by rock3r** [comment]
I have added a meaningful description for this PR. I'm working on adding more tests, but this PR is not blocked by that. They can be added in a later PR.

@igordmn the iOS build is failing due to excessive egress usage, trying to download Skia releases. This should be fixed by using the Gradle caching in the CI jobs, which I have added, but it looks like the `downloadSkia*` tasks are doing downloads in a non-standard way that doesn't get cached. This will likely need to be fixed by either explicitly caching these downloaded files for CI jobs, or changing those download tasks to actual dependencies which will get properly cached.

I have also noticed that `SkikoProperties#isCIBuild` is looking for TeamCity, but that'll never be true on GH Actions. It should be changed to look for the "CI" property.

I can tackle these things in a separate PR once this is shipped.

**Comment by rock3r** [comment]
The CI is red on iOS because of excessive downloading (see above comment), and Linux is failing for reasons I can't understand:

```
Execution failed for task ':skiko:compileKotlinAndroid'.
> Could not resolve all files for configuration ':skiko:androidCompileClasspath'.
   > For input string: "33-ext4"
```

@igordmn any advice on fixing would be welcome.

File: skiko/src/awtTest/kotlin/org/jetbrains/skiko/SkiaSystemFontProviderTest.kt
Comment (igordmn) [reply skiko/src/awtTest/kotlin/org/jetbrains/skiko/SkiaSystemFontProviderTest.kt:38]: Ubuntu:
```
java.lang.AssertionError: These AWT font families are missing:
 * Droid Sans
 * Droid Sans Mono
 * Droid Sans Mono Dotted
 * Droid Sans Mono Slashed
 * Droid Serif
 * Fira Code
 * Inconsolata
 * Inter
 * JetBrains Mono
 * Roboto
 * Source Code Pro
```
They are also in the JBR folder

Comment (igordmn) [reply skiko/src/awtTest/kotlin/org/jetbrains/skiko/SkiaSystemFontProviderTest.kt:38]: MacOs Monterey 12.5., JBR17:
```
0 = ".SF NS"
1 = ".SF NS Mono"
2 = "Athelas"
3 = "Courier"
4 = "Droid Sans"
5 = "Droid Sans Mono"
6 = "Droid Sans Mono Dotted"
7 = "Droid Sans Mono Slashed"
8 = "Droid Serif"
9 = "Fira Code"
10 = "Hiragino Kaku Gothic Pro"
...
15 = "Hiragino Mincho Pro"
16 = "Inconsolata"
17 = "Inter"
18 = "Iowan Old Style"
19 = "JetBrains Mono"
20 = "Marion"
21 = "Noto Sans Adlam"
...
122 = "Noto Serif Yezidi"
123 = "Roboto Light"
124 = "Roboto Thin"
125 = "Seravek"
126 = "Source Code Pro"
127 = "Superclarendon"
128 = "Times"
```
Those are in the JBR folder, in /System/Library/Fonts/Supplementa, and in /System/Library/Fonts/Supplemental.

Don't know what happens, but usual mac applications (Pages, TextEdit, Chrome) also don't see fonts from /System/Library/Fonts.

File: .github/workflows/ci.yml
Comment (rock3r) [reply .github/workflows/ci.yml:49]: The only note is this doesn't publish the reports when a PR originates from a fork due to limitations in the token scope (by design, for security).

File: skiko/src/awtTest/kotlin/org/jetbrains/skiko/SkiaSystemFontProviderTest.kt
Comment (rock3r) [reply skiko/src/awtTest/kotlin/org/jetbrains/skiko/SkiaSystemFontProviderTest.kt:38]: Hmmm I haven't run the tests on JBR I suppose, but I can look into this. JDK classpath fonts are not a big deal, as this seems another JBR-only issue (the tests pass for me when not using JBR, and on the CI), since JBR embeds some fonts. I'll look into loading those fonts into the list too, but there is a concern, in the sense that non-system fonts aren't accessible to Skia automatically. What I was thinking is adding a "if you see fonts in the JDK, add them to the list of custom fonts so they can be used normally" — they are, after all, not system fonts.

As for the reverse, Skia having fonts that AWT doesn't know about, that's ignored by design. AWT can't deal with a number of font types (variable fonts, composite fonts, COLRv1 emoji fonts, etc) since its implementation is stuck to 20+ years ago, so it's only normal Skia knows about more fonts than AWT. For the AWT -> Skia interop this isn't a problem as long as we guarantee all AWT fonts are available to Skia (once the JDK classpath fonts loading is added). For the Skia -> AWT interop there isn't much we can do, but users can either rely on AWT's silent font substitution mechanism, or manually check the available AWT font families and decide what to do if missing.

File: skiko/src/awtMain/kotlin/org/jetbrains/skiko/AwtFontManager.kt
Comment (rock3r) [reply skiko/src/awtMain/kotlin/org/jetbrains/skiko/AwtFontManager.kt:24]: Not sure what you are referring to?

File: skiko/build.gradle.kts
Comment (rock3r) [reply skiko/build.gradle.kts:295]: Ah, when I checked ByteBuddy I missed this bit. That's unfortunate.

IDEA builds do have the open directives we need, so we don't need BB there. On the other hand, this would mean we need to document this need for all other scenarios. This meand that any CfD application that runs on the JBR will not work unless they add the opens directive (you can't easily figure out whether the opens directive is present). I can look into adding an agent to inject the directives without using ByteBuddy: it should be possible, as BB doesn't use any provate APIs or particular black magic. It'll just take more code.

File: skiko/src/awtMain/kotlin/org/jetbrains/skiko/Convertors.jvm.kt
Comment (rock3r) [reply skiko/src/awtMain/kotlin/org/jetbrains/skiko/Convertors.jvm.kt]: Ah, I didn't change this on purpose. It must've been done when I ran Cleanup Code on the file. Reverted in 6095fe13163dfc4bd0ad20aa0f9c92adbdf0b91a

File: skiko/src/awtMain/kotlin/org/jetbrains/skiko/InternalSunApiOpener.kt
Comment (rock3r) [reply skiko/src/awtMain/kotlin/org/jetbrains/skiko/InternalSunApiOpener.kt]: No, I had just noticed it and removed it in 4fe1353e015dfa9378bd3370c9f73c86431a5cd7. Thanks :)

File: skiko/src/awtMain/kotlin/org/jetbrains/skiko/InternalSunApiOpener.kt
Comment (rock3r) [reply skiko/src/awtMain/kotlin/org/jetbrains/skiko/InternalSunApiOpener.kt]: Nope, same as previous comment. It was supposed to be another one, which I forgot to change it to. But it's fixed now together with the previous one. I have also improved this method by adding additional checks and tests for them (indirectly, via `AwtFontManager`).

File: skiko/src/awtTest/kotlin/org/jetbrains/skiko/AwtFontManagerTest.kt
Comment (rock3r) [reply skiko/src/awtTest/kotlin/org/jetbrains/skiko/AwtFontManagerTest.kt]: Done in eda02804990da6d6eabd3393401de0549da25a8c

File: skiko/src/awtMain/kotlin/org/jetbrains/skiko/Convertors.jvm.kt
Comment (rock3r) [reply skiko/src/awtMain/kotlin/org/jetbrains/skiko/Convertors.jvm.kt:268]: Done in eda02804990da6d6eabd3393401de0549da25a8c

File: skiko/.run/awtTest.run.xml
Comment (rock3r) [reply skiko/.run/awtTest.run.xml]: Done in 7daf8b2fd72615467be16101d9661c4a592dd68a

File: .github/workflows/ci.yml
Comment (rock3r) [reply .github/workflows/ci.yml:28]: Fixed as per previous comment.

File: skiko/build.gradle.kts
Comment (rock3r) [reply skiko/build.gradle.kts]: I bumped the Java toolchain version to 17 because there was code around that would be marked red in the IDE since it was using newer APIs and the Gradle plugin wasn't able to pick up the correct language level (it would always reset to 6 when syncing).

But no there is no hard requirement on 17 — it should be 11, as I haven't changed anything else. I'm fixing this, sorry. Must have been when I was testing something out and I forgot to switch it back to 11.

Done in 7daf8b2fd72615467be16101d9661c4a592dd68a

File: skiko/src/jvmTest/kotlin/org/jetbrains/skiko/TaskTest.kt
Comment (rock3r) [reply skiko/src/jvmTest/kotlin/org/jetbrains/skiko/TaskTest.kt:188]: Fixed in 3e990b7595666aa700d0c368951026b121c71a28

File: skiko/src/awtMain/kotlin/org/jetbrains/skiko/AwtFontManager.kt
Comment (igordmn) [reply skiko/src/awtMain/kotlin/org/jetbrains/skiko/AwtFontManager.kt:24]: replace all public functions such as:
```
fun customFamilyNames() = customTypefaceCache.familyNames().toSortedSet()
```
to
```
fun customFamilyNames(): Set<String> = customTypefaceCache.familyNames().toSortedSet()
```

**Comment by rock3r** [comment]
@igordmn any idea what this issue on the Linux CI is?

```
Execution failed for task ':skiko:compileKotlinAndroid'.
> Could not resolve all files for configuration ':skiko:androidCompileClasspath'.
   > For input string: "33-ext4"
```

**Comment by igordmn** [comment]
>  > For input string: "33-ext4"

It happens in all our [builds](https://github.com/JetBrains/skiko/actions/runs/3989735000/jobs/6842525860), just ignore it, we'll fix it separately

**Comment by rock3r** [comment]
@igordmn ready to ship! With big thanks to @dima-avdeev-jb for the fixes I rebased onto :)

**Comment by igordmn** [comment]
Awesome! I will review it soon.

**Comment by jimgoog** [comment]
Ping @igordmn

File: skiko/src/awtTest/kotlin/org/jetbrains/skiko/SkiaSystemFontProviderTest.kt
Comment (igordmn) [reply skiko/src/awtTest/kotlin/org/jetbrains/skiko/SkiaSystemFontProviderTest.kt:38]: We still have this issue. What is the decision to resolve it?

We have fonts in the JDK folder, and have some fonts `/System/Library` on macOS. How do we register them? If there is a way to load them, we need to use it in tests, so they don't fail. If there is no way, we need to find another solution (maybe even fillter fonts in the IDEA settings itself)

File: skiko/build.gradle.kts
Comment (igordmn) [inline skiko/build.gradle.kts:222]: Does this property override our JDK that we set in the Project settings?

If this is enabled, I have java.vendor = Eclipse Adoptium, java.version = 11.0.18, so our test don't run.

Probably worth to write a comment for the JBR test, that we need to comment this property first.

File: skiko/docker/linux-android-amd64/Dockerfile
Comment (rock3r) [reply skiko/docker/linux-android-amd64/Dockerfile:8]: This needs reverting, was left over from when we had bumped the JDK version

File: skiko/build.gradle.kts
Comment (rock3r) [reply skiko/build.gradle.kts:222]: We can revert this I think, was more useful when we had bumped the JDK. It shouldn't override anything though, in a Gradle project the IDE project settings are derived from Gradle.

File: skiko/build.gradle.kts
Comment (rock3r) [reply skiko/build.gradle.kts:294]: I totally forgot this needed doing, sorry 😔

I need to find some time to create the agent without ByteBuddy and see if it works, but a starting point would be to just attempt accessing the APIs and erroring out if they aren't accessible. Do you think it's reasonable asking folks to add an opens directive to the command line to run their projects? I can have a very explicit explanation in the error message.

**Comment by rock3r** [comment]
Well I suppose the agents route is officially dead. https://mail.openjdk.org/pipermail/serviceability-dev/2023-March/046868.html

We'll have to instruct folks to add the relevant opens directives to their command line args. I'll remove all that logic and just put a println to stderr that warns folks about this need if the opens are missing.

**Comment by igordmn** [comment]
> We'll have to instruct folks to add the relevant opens directives to their command line args

Looks okay. The main client is IDEA Platform anyway, where it should be already done?

File: skiko/build.gradle.kts
Comment (rock3r) [reply skiko/build.gradle.kts]: Removed the ByteBuddy dependency and code

Comment (rock3r) [reply skiko/build.gradle.kts]: It was removed in a52b316549b770d7998e7af940919750d725d542

**Comment by rock3r** [comment]
@igordmn I just pushed a few things — it's not all done yet but it's a step forward.

* ByteBuddy is gone, and we show a message if the opens directive is missing
* I have started working on adding the ability to load embedded fonts, BUT
* I can't test it cause Gradle doesn't seem to pass the JVM args I set, which are needed to make the tests work

I would need help with the latter issue, if you know what's up

**Comment by rock3r** [comment]
Oh and, I expect (hope!) that at least some tests will fail in this in-between state

**Comment by rock3r** [comment]
@igordmn I think I am done addressing all the issues. I have spent a non-trivial amount of time chasing random edge cases that cropped up over time, and it'd be great if someone from the JBR team can come and check if the special casing code I wrote makes any sense.

Besides, there's one test I can't get to run: `SkiaFontProviderTest#should be able to resolve logical fonts when running on JetBrains Runtime()`. It fails because it can't detect the `java.desktop/sun.font` module as being opened, but as far as I can tell, it is: the Gradle task has the correct parameters (checked at runtime as well) and if I attach the debugger, then it's suddenly passing the check. I have no idea what's going on here; if you do, or can loop in someone that does, that would be great :)

I also rebased onto the main branch and now it looks like the CI is also failing with errors in the build script, in places I didn't touch.

**Comment by igordmn** [comment]
Thanks! Will look at it in the near days

File: skiko/src/awtTest/kotlin/org/jetbrains/skiko/SkiaFontProviderTest.kt
Comment (igordmn) [reply skiko/src/awtTest/kotlin/org/jetbrains/skiko/SkiaFontProviderTest.kt:66]: It requires a lot of rewriting, so we can just keep in mind that Compose panels won't support variable fonts (maybe only some of them). But that can be a blocker for production ready functionality.

File: skiko/src/awtMain/kotlin/org/jetbrains/skiko/InternalSunApiChecker.kt
Comment (rock3r) [reply skiko/src/awtMain/kotlin/org/jetbrains/skiko/InternalSunApiChecker.kt:76]: I'll remove the recommendation from here 👍

File: skiko/src/awtTest/kotlin/org/jetbrains/skiko/SkiaFontProviderTest.kt
Comment (rock3r) [reply skiko/src/awtTest/kotlin/org/jetbrains/skiko/SkiaFontProviderTest.kt:66]: I'd punt variable fonts support for a subsequent PR — we need to get the basic support shipped ASAP. This may be a blocker down the road but isn't yet for the day 1 scenarios we're supporting, and this PR is already a huge step forward from the previous situation.

Nikolay said he was going to take on further changes to the font loading code, I can work with him to make those changes. What do you think?

File: skiko/src/awtMain/kotlin/org/jetbrains/skiko/AwtFontUtils.kt
Comment (igordmn) [reply skiko/src/awtMain/kotlin/org/jetbrains/skiko/AwtFontUtils.kt]: Could you correct the code in your branch after the fix? GitHub can't remove other lines below.

File: skiko/src/jvmTest/kotlin/org/jetbrains/skiko/TaskTest.kt
Comment (igordmn) [inline skiko/src/jvmTest/kotlin/org/jetbrains/skiko/TaskTest.kt:59]: I made [a PR with refactorings](https://github.com/rock3r/skiko/pull/1). Please, review it.

One more thing I wanted to do - to check it on Linux. Will check today or tomorrow.

After that we can merge the PR.

**Review by igordmn — COMMENTED** [review]
--

File: skiko/src/jvmTest/kotlin/org/jetbrains/skiko/TaskTest.kt
Comment (rock3r) [reply skiko/src/jvmTest/kotlin/org/jetbrains/skiko/TaskTest.kt:59]: Thanks Igor! I did review it, only a couple minor things. Will update the PR description after those are addressed and that PR merged.

Comment (rock3r) [reply skiko/src/jvmTest/kotlin/org/jetbrains/skiko/TaskTest.kt]: And done, PR description updated. Added a note about embedded JVM fonts and removed outdated statements

**Comment by igordmn** [comment]
> @igordmn looks like we're ready to 🚢 ?

If you can run tests on Linux+JBR again, we can merge it now. Otherwise I would like to wait for tomorrow's evening, when I able to do that.

**Comment by rock3r** [comment]
I won't be able to set up a Linux virtual env before Monday, the one I used got thrashed and I had to dump it. It was not really working well in any case :(

**Review by igordmn — COMMENTED** [review]
I did [the last checks and fixes](https://github.com/rock3r/skiko/pull/2). We can merge the PR after merging them 🎉 .

Also extracted [refactorings](https://github.com/JetBrains/skiko/pull/729), please review.

File: skiko/src/awtTest/kotlin/org/jetbrains/skiko/SkiaFontProviderTest.kt
Comment (igordmn) [reply skiko/src/awtTest/kotlin/org/jetbrains/skiko/SkiaFontProviderTest.kt]: > On Windows/Linux we can retrieve the file name, and load it in Skia.

On Windows we have to get the native pointer too with this approach :(

Both "Cambria" and "Cambria Math" variant point to "C:\Windows\Fonts\cambria.ttc"

Comment (rock3r) [reply skiko/src/awtTest/kotlin/org/jetbrains/skiko/SkiaFontProviderTest.kt]: Ah that's because that is a composite font (`.ttc` file). It's something we can't do too much about, as they are only kinda-sorta supported by AWT. What AWT does when it loads ttc files is it goes "oh, cool, get me all the fonts in this file" and then does `fonts[0]`, with a nice TODO saying essentially "huh we may wanna revisit this"

**Review by igordmn — APPROVED** [review]

**Comment by rock3r** [comment]
Wooooo we did it! 🎉 Thanks Igor 🙏

**Pull request merged**
